Changes in module 'nf-core/vcftools'
--- modules/nf-core/vcftools/main.nf
+++ modules/nf-core/vcftools/main.nf
@@ -91,10 +91,15 @@
 
     def bed_arg  = (args.contains('--bed')) ? "--bed ${bed}" :
         (args.contains('--exclude-bed')) ? "--exclude-bed ${bed}" :
-        (args.contains('--hapcount')) ? "--hapcount ${bed}" : ''
+        (args.contains('--hapcount')) ? "--hapcount ${bed}" :
+        (args.contains('--positions')) ? "--positions ${bed}" :
+        (args.contains('--exclude-positions')) ? "--exclude-positions ${bed}"  : ''
+
     args_list.removeIf { it.contains('--bed') }
     args_list.removeIf { it.contains('--exclude-bed') }
     args_list.removeIf { it.contains('--hapcount') }
+    args_list.removeIf { it.contains('--positions') }
+    args_list.removeIf { it.contains('--exclude-positions') }
 
     def diff_variant_arg = (args.contains('--diff')) ? "--diff ${diff_variant_file}" :
         (args.contains('--gzdiff')) ? "--gzdiff ${diff_variant_file}" :

************************************************************
