Changes in module 'nf-core/samtools/merge'
--- modules/nf-core/samtools/merge/meta.yml
+++ modules/nf-core/samtools/merge/meta.yml
@@ -43,6 +43,10 @@
       type: file
       description: Index of the reference file the CRAM was created with (optional)
       pattern: "*.fai"
+  - gzi:
+      type: file
+      description: Index of the reference file the CRAM was created with (optional)
+      pattern: "*.gzi"
 output:
   - meta:
       type: map
@@ -65,6 +69,10 @@
       type: file
       description: BAM index file (optional)
       pattern: "*.csi"
+  - crai:
+      type: file
+      description: CRAM index file (optional)
+      pattern: "*.crai"
 authors:
   - "@drpatelh"
   - "@yuukiiwa "

--- modules/nf-core/samtools/merge/main.nf
+++ modules/nf-core/samtools/merge/main.nf
@@ -11,11 +11,13 @@
     tuple val(meta), path(input_files, stageAs: "?/*")
     tuple val(meta2), path(fasta)
     tuple val(meta3), path(fai)
+    tuple val(meta4), path(gzi)
 
     output:
     tuple val(meta), path("${prefix}.bam") , optional:true, emit: bam
     tuple val(meta), path("${prefix}.cram"), optional:true, emit: cram
     tuple val(meta), path("*.csi")         , optional:true, emit: csi
+    tuple val(meta), path("*.crai")        , optional:true, emit: crai
     path  "versions.yml"                                  , emit: versions
 
 
@@ -31,6 +33,7 @@
     samtools \\
         merge \\
         --threads ${task.cpus-1} \\
+        --write-index \\
         $args \\
         ${reference} \\
         ${prefix}.${file_type} \\

************************************************************
