# sanger-tol/variantcalling: Output

## Introduction

This document describes the output produced by the pipeline.

The directories listed below will be created in the results directory after the pipeline has finished. All paths are relative to the top-level results directory.

## Pipeline overview

The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes data using the following steps:

- [Readmapping Alignments](#readmapping-alignments) - Optional aligned CRAM files generated by minimap2
- [Alignments Statistics](#alignments-statistics) Optional statistics files generated by samtools
- [VCFtools Processing](#vcftools-processing) Heterozygosity and per site nucleotide diversity calculated by VCFtools
- [PacBio Variant Calling](#pacbio-variant-calling) - VCF and GVCF compressed files generated by DeepVariant
- [Pipeline information](#pipeline-information) - Report metrics generated during the workflow execution

### Readmapping Alignments

The unaligned PacBio read data is being filtered and aligned using `minimap2`. The CRAM files from the same sample will be merged.

<details markdown="1">
<summary>Output files</summary>

- `readmapping`
  - Aligned CRAM files: `<fasta_name>.pacbio.<sample_name>.cram`.
  - Aligned CRAM index files: `<fasta_name>.pacbio.<sample_name>.cram.crai`.

</details>

### Alignments Statistics

The statistics for the aligned CRAM files will be calculated using `samtools`.

<details markdown="1">
<summary>Output files</summary>

- `statistics`
  - Comprehensive statistics from alignment file: `<fasta_name>.pacbio.<sample_name>.stats`.
  - Number of alignments for each FLAG type: `<fasta_name>.pacbio.<sample_name>.flagstats`.
  - Alignment summary statistics: `<fasta_name>.pacbio.<sample_name>.idxstats`.

</details>

### VCFtools Processing

<details markdown="1">
<summary>Output files</summary>

- Heterozygosity generated by VCFtools: `<fasta_name>.pacbio.<sample_name>_deepvariant.vcf.het`.
- Per site nucleotide diversity calculated by VCFtools: `<fasta_name>.pacbio.<sample_name>_deepvariant.vcf.sites.pi`.

</details>

### PacBio Variant Calling

The aligned PacBio read data is used to call variants with DeepVariant. This is done by splitting the genome fasta file for speed efficiency. `BCFTOOLS` is used to combine the split `VCF` and `GVCF` files generated by `DEEPVARIANT`.

<details markdown="1">
<summary>Output files</summary>

- `variant_calling`
  - Compressed VCF files: `<fasta_name>.pacbio.<sample_name>_deepvariant.vcf.gz`.
  - Index of compressed VCF files: `<fasta_name>.pacbio.<sample_name>_deepvariant.vcf.gz.[tbi|csi]`.
  - Compressed GVCF files: `<fasta_name>.pacbio.<sample_name>_deepvariant.g.vcf.gz`.
  - Index of compressed GVCF files: `<fasta_name>.pacbio.<sample_name>_deepvariant.g.vcf.gz.[tbi|csi]`.
  - `reports`
    - HTML files: `<fasta_name>.pacbio.<sample_name>_[vcf|gvcf].stats.visual_report.html`.

</details>

### Pipeline information

<details markdown="1">
<summary>Output files</summary>

- `pipeline_info/variantcalling/`
  - Reports generated by Nextflow: `execution_report.html`, `execution_timeline.html`, `execution_trace.txt` and `pipeline_dag.dot`/`pipeline_dag.svg`.
  - Reports generated by the pipeline: `pipeline_report.html`, `pipeline_report.txt` and `software_versions.yml`. The `pipeline_report*` files will only be present if the `--email` / `--email_on_fail` parameter's are used when running the pipeline.
  - Reformatted samplesheet files used as input to the pipeline: `samplesheet.valid.csv`.
  - Parameters used by the pipeline run: `params.json`.

</details>

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.
